sqlx-mysql/src/connection/infile.rs

```diff
@@ -0,0 +1,298 @@
+            if self.buf.len() >= MAX_MYSQL_PACKET_SIZE + 4 {
+                assert_eq!(self.buf.len(), MAX_MYSQL_PACKET_SIZE + 4);
+                self.drain_packet(MAX_MYSQL_PACKET_SIZE).await?;
+                assert!(self.buf.is_empty());
+                self.buf.extend_from_slice(&[0; 4]);
+            }
+            right = right_;
+        }
+        Ok(())
+    }
+
+    /// Flush the stream.
+    pub async fn flush(&mut self) -> Result<(), Error> {
+        if self.buf.len() > 4 {
+            // Cannot have multiple packets in buffer, as they would have been drained by write() already
+            assert!(self.buf.len() <= MAX_MYSQL_PACKET_SIZE + 4);
+            self.drain_packet(self.buf.len() - 4).await?;
+        }
+        Ok(())
+    }
+
+    async fn drain_packet(&mut self, len: usize) -> Result<(), Error> {
+        self.buf[0..3].copy_from_slice(&(len as u32).to_le_bytes()[..3]);
+        self.buf[3] = self.conn.stream.sequence_id;
+        self.conn
+            .stream
+            .socket
+            .socket_mut()
+            .write(&self.buf[..len + 4])
+            .await?;
+        self.buf.drain(..len + 4);
+        self.conn.stream.sequence_id = self.conn.stream.sequence_id.wrapping_add(1);
+        Ok(())
+    }
+
+    /// Finish sending the LOCAL INFILE data to the server.
+    ///
+    /// This must always be called after you're done writing the data.
+    /// Returns the number of rows that were inserted.
+    pub async fn finish(mut self) -> Result<u64, Error> {
+        self.flush().await?;
+        self.conn.stream.send_empty_response().await?;
+        let packet = self.conn.stream.recv_packet().await?;
+        let packet = packet.ok()?;
+        Ok(packet.affected_rows)
+    }
+}
+
+/// A writer that writes to a [`MySqlLocalInfile`] stream.
+pub struct InfileWriter<'a> {
+    stream: &'a mut MySqlStream,
+    send: Option<SendPacket>,
+}
+
+impl<'a> InfileWriter<'a> {
+    fn new(stream: &'a mut MySqlStream) -> Self {
+        Self { stream, send: None }
+    }
+}
+
+impl<'a> AsyncWrite for InfileWriter<'a> {
+    fn poll_write(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<futures_io::Result<usize>> {
+        if self.send.is_some() {
+            match self.as_mut().poll_flush(cx) {
+                Poll::Ready(Ok(())) => {
+                    self.send = None;
+                }
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+        // If the code reaches here the flush was succesful
+        assert!(self.send.is_none());
+        let mut send = SendPacket::new(buf, self.stream.sequence_id);
+        self.stream.sequence_id = self.stream.sequence_id.wrapping_add(1);
+        // Try to poll the send future right now
+        match Pin::new(&mut send).poll_send(cx, self.stream.socket_mut()) {
```

Comment:
Why are you bypassing the connection's internal buffer?

Reply:
My idea was that you either use the "easy" way of calling `MySqlLocalInfile::send` a couple times, which does buffering, or you want to do a different style of buffering so you need a "direct" API. In my own usage, I wrap the "direct" writer with my own BufWriter to provide buffering of infile packets.

Reply:
My thought is, we already have a perfectly buffer in the connection. Why force the user to allocate another?

Reply:
@jelle-bigbridge Do you plan to get back to working on this PR? This feature would be very useful for us, if you're busy I could pick it up.

Reply:
I need feedback from the maintainers to finish this to be quite honest. If this direct mode needs to be reworked I can do it, but I need some direction to take this in. I don't want to go and make changes myself without knowing if it will help to get it merged

Reply:
@pingiun my feedback was to not bypass the connection's internal buffer.

Reply:
For copying from an `AsyncRead` type, you can include a `read_from` method similar to `PgCopyIn`: https://docs.rs/sqlx/latest/sqlx/postgres/struct.PgCopyIn.html#method.read_from

The trait situation isn't great there, fixing it depends on having async I/O traits in `std`, traits which I was told were coming more than two years ago and still haven't materialized: https://github.com/launchbadge/sqlx/issues/1669#issuecomment-1028914973